arch/xtensa/core/xtensa-win0.S

```diff
@@ -0,0 +1,523 @@
+	s32i a10, a1, __xtensa_win0_ctx_t_a10_OFFSET
+	s32i a11, a1, __xtensa_win0_ctx_t_a11_OFFSET
+	s32i a12, a1, __xtensa_win0_ctx_t_a12_OFFSET
+	s32i a13, a1, __xtensa_win0_ctx_t_a13_OFFSET
+	s32i a14, a1, __xtensa_win0_ctx_t_a14_OFFSET
+	s32i a15, a1, __xtensa_win0_ctx_t_a15_OFFSET
+
+	mov a2, a1 /* context is argument */
+	callx0 a0
+	rsil a0, 0xf
+	rsr a2, ZSR_CPU
+	MOD_NEST_COUNT -1 a2 a0
+	mov a3, a1
+	j restore
+
+/* System call exceptions are an extremely pruned case of general
+ * exception handling:
+ *
+ * + We know we were a trap from a userspace thread context, and don't
+ *   need to check for nested interrupts or entry from kernel threads
+ * + We don't need to do a full context switch: userspace knew it was
+ *   trapping and will have pre-spilled state we don't need to save.
+ *   (In practice we preserve a0-1 and a12-15, and leave a return value
+ *   in a2)
+ * + Likewise we know userspace isn't in the middle of code using
+ *   compiler-generated state registers like
+ *   SAR/SCOMPARE/loops/etc... (The compiler won't emit a LOOP
+ *   construction if it can't prove there is no nesting, so a simple
+ *   __asm__ block defeats it)
+ * + The parameter registers are the same as the C ABI, so we can
+ *   call the C handlers directly out of the table.  This will
+ *   preserve a12-15 for us, so we don't have to spill them.
+ *
+ * With a few special conventions and rules:
+ *
+ * + The syscall number itself is passed in a11 and we use a10 as
+ *   scratch (allows space for 8 parameters in a2-a9)
+ * + We stash the userspace a0/a1/EPC in a special location in the
+ *   thread struct, as the thread's own context now belongs to this
+ *   syscall (which can itself be switched out!)
+ * + We must switch to the thread's priviledged stack
+ * + The nine unused-at-exit registers (a3-a11) must be wiped on exit
+ *   such that they don't contain stale kernel data.
+ * + We still have to unmask EXCM, set RING to zero, and fix INTLEVEL
+ *   due to a bug on some platforms that leaves it at zero incorrectly
+ * + And we return with RFI, not RFE, because we have to set a PS with
+ *   RING re-elevated
+ */
+syscall_handler:
+	rsr a10, ZSR_CPU
+	l32i a10, a10, ___cpu_t_current_OFFSET
+	rsr a0, ZSR_A0SAVE
+	s32i a0, a10, ___thread_t_arch_OFFSET + __xtensa_win0_ctx_t_user_a0_OFFSET
+	s32i a1, a10, ___thread_t_arch_OFFSET + __xtensa_win0_ctx_t_user_a1_OFFSET
+	rsr a0, EPC1
+	addi a0, a0, 3 /* skip over syscall instruction */
+	s32i a0, a10, ___thread_t_arch_OFFSET + __xtensa_win0_ctx_t_user_pc_OFFSET
+	rsr a0, PS
+	addi a0, a0, -(PS_EXCM_MASK)
+	s32i a0, a10, ___thread_t_arch_OFFSET + __xtensa_win0_ctx_t_user_ps_OFFSET
+
+	/* Further cook PS to set RING=0 and UM=0, then set our sp and unmask */
+	movi a1, ~(PS_RING_MASK | PS_UM_MASK)
+	and a0, a0, a1
+#ifdef CONFIG_USERSPACE
+	l32i a1, a10, ___thread_t_arch_OFFSET + ___thread_arch_t_priv_sp_OFFSET
+#else
+	movi a1, _mock_priv_stack
+	l32i a1, a1, 0
+#endif
+	wsr a0, PS
+
```

Comment:
assuming that you have to re-enable interruptions / exceptions before jump into the handler
you can be switched after setting `PS`  and  syscall parameters are not saved in this context stack ...

Reply:
The interrupted userspace a0/a1/pc/ps values (the ones we need to preserve) are already stashed in the thread struct at this point.  All a context switch arriving now has to do is save our kernel space GPRs.

Reply:
that happens only with interrutptions ritght ? inside `must_switch`

I remember having seen case where a syscall triggers a context switch  without an interruption happens and AFAIU `xtensa_switch` does not save them.